rpc/types/src/macros.rs

```diff
@@ -0,0 +1,242 @@
+//! Macros.
+
+//---------------------------------------------------------------------------------------------------- Struct definition
+/// A template for generating 2 `struct`s with a bunch of information filled out.
+///
+/// These are the RPC request and response `struct`s.
+///
+/// It's best to see the output of this macro via the documentation
+/// of the generated structs via `cargo doc`s to see which parts
+/// generate which docs.
+///
+/// See the [`crate::json`] module for example usage.
+///
+/// # Macro internals
+/// This macro has 2 branches with almost the same output:
+/// 1. An empty `Request` type
+/// 2. An `Request` type with fields
+///
+/// The first branch is the same as the second with the exception
+/// that if the caller of this macro provides no fields, it will
+/// generate:
+/// ```
+/// pub type Request = ();
+/// ```
+/// instead of:
+/// ```
+/// pub struct Request {/* fields */}
+/// ```
+///
+/// This is because having a bunch of types that are all empty structs
+/// means they are not compatible and it makes it cumbersome for end-users.
+/// Really, they semantically are empty types, so `()` is used.
+///
+/// Again, other than this, the 2 branches do (should) not differ.
+///
+/// FIXME: there's probably a less painful way to branch here on input
+/// without having to duplicate 80% of the macro. Sub-macros were attempted
+/// but they ended up unreadable. So for now, make sure to fix the other
+/// branch as well when making changes. The only de-duplicated part is
+/// the doc generation with [`define_request_and_response_doc`].
+macro_rules! define_request_and_response {
```

Comment:
There's a lot of `Request` types that have no inputs; generating a bunch of incompatible `struct` types that are actually all the same type (`()`) would be unwieldy for callers, e.g. the interface code would have to reference specific types even though they are all `()`.

So, with this input: `Request {}`, the macro now generates this:

![image](https://github.com/Cuprate/cuprate/assets/101352116/c21b02e4-df9d-44b0-b00b-114a0f48b686)

Now callers can just use `()`. A downside is that the macro has some duplicate code within the 2 branches, I don't think it's too bad though.

Reply:
Is this macro going to be generaing the binary types as well? 

If so you will need to add `epee_encoding`